File: examples/bevholt/bh.R
# bh.R
# Beverton-Holt Model R                = aS/(1+bS)*exp(-epsilon*sigR)
library(TMB)

data <- read.table("bevholt.dat",header=TRUE)
data <- list(R    =data$R,S=data$S)
pars <- list(log_a= log(2.1), log_b = 0, log_sig = 0, log_tau = log(0.1),
		 u        = rep(0,length=length(data$S)))

compile("bh.cpp")
dyn.load(dynlib("bh"))

# f   <- MakeADFun(data,parameters=pars,random=c("u"),ADreport = FALSE)
f   <- MakeADFun(data,parameters=pars,map=list(log_tau=factor(c(NA))))
fit <- nlminb(f$par,f$fn,f$gr) 



mle <- f$env$last.par.best
rep <- sdreport(f)

print(mle)
print(summary(rep))

plot(data$S,data$R,pch=20,col=1)
points(f$report(),pch =20,col=2)

File: examples/bevholt/bh.cpp
// bh.cpp
#include <TMB.hpp>

template <class Type>
Type objective_function<Type>::operator()()
{
	// DATA
	DATA_VECTOR(R)
	DATA_VECTOR(S)

	// PARAMETERS
	PARAMETER(log_a);
	PARAMETER(log_b);
	PARAMETER(log_sig);
	PARAMETER(log_tau);
	PARAMETER_VECTOR(u);

	Type a = exp(log_a);
	Type b = exp(log_b);
	Type sig = exp(log_sig);
	Type tau = exp(log_tau);

	vector<Type>    x = S * exp(u);
	vector<Type>    y = a * x / ( Type(1.0) + b * x );
	vector<Type> epsi = log(R) - log(y);
	Type nll = 0;
	nll -= sum(dnorm(epsi,Type(0.0),sig,true));
	nll -= sum(dnorm(u,Type(0.0),tau,true));


	REPORT(x);
	REPORT(y);
	REPORT(epsi);
	ADREPORT(y);

	return nll;
}

